// yahoo-finance2/api/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FailedYahooValidationError = exports.NoEnvironmentError = exports.InvalidOptionsError = exports.HTTPError = exports.BadRequestError = void 0;
class BadRequestError extends Error {
    constructor() {
        super(...arguments);
        this.name = "BadRequestError";
    }
}
exports.BadRequestError = BadRequestError;
class HTTPError extends Error {
    constructor() {
        super(...arguments);
        this.name = "HTTPError";
    }
}
exports.HTTPError = HTTPError;
class InvalidOptionsError extends Error {
    constructor() {
        super(...arguments);
        this.name = "InvalidOptionsError";
    }
}
exports.InvalidOptionsError = InvalidOptionsError;
class NoEnvironmentError extends Error {
    constructor() {
        super(...arguments);
        this.name = "NoEnvironmentError";
    }
}
exports.NoEnvironmentError = NoEnvironmentError;
class FailedYahooValidationError extends Error {
    constructor(message, { result, errors }) {
        super(message);
        this.name = "FailedYahooValidationError";
        this.result = result;
        this.errors = errors;
    }
}
exports.FailedYahooValidationError = FailedYahooValidationError;
