// yahoo-finance2/api/modules/quote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const queryOptionsDefaults = {};
function quote(query, queryOptionsOverrides, moduleOptions) {
    const symbols = typeof query === "string" ? query : query.join(",");
    return this._moduleExec({
        moduleName: "quote",
        query: {
            url: "https://query2.finance.yahoo.com/v7/finance/quote",
            schemaKey: "#/definitions/QuoteOptions",
            defaults: queryOptionsDefaults,
            runtime: { symbols },
            overrides: queryOptionsOverrides,
            transformWith(queryOptions) {
                // Options validation ensures this is a string[]
                if (queryOptions.fields)
                    queryOptions.fields.join(",");
                return queryOptions;
            },
        },
        result: {
            schemaKey: "#/definitions/QuoteResponse",
            transformWith(result) {
                if (!result.quoteResponse)
                    throw new Error("Unexpected result: " + JSON.stringify(result));
                return result.quoteResponse.result;
            },
        },
        moduleOptions,
    }).then((results) => {
        return typeof query === "string"
            ? results[0]
            : results;
    });
}
exports.default = quote;
